capstone/src/main/webapp/script.js

```diff
@@ -31,14 +32,160 @@ async function submitInput() {
   if (!langElement) {
     return;
   }
+
+  // Get the selected attributes
   const attributes = document.getElementById("available-attributes").getElementsByTagName('input');
-  const requestedAttributes = []
+  const requestedAttributes = [];
   for (const attribute of attributes) {
     if (attribute.checked) {
       requestedAttributes.push(attribute.value);	
     }	
   }
-  await callPerspective(textElement.value, langElement.value, requestedAttributes);
+
+  // Get the selected analysis type
+  document.getElementById('analysis-container').innerHTML = '';
+  const radios = document.getElementsByName('analysisRadios');
+  var tokenizer = "";
+  for (i = 0; i < radios.length; i++) {
+    if (radios[i].checked) {
+      tokenizer = radios[i].value;
+      break;
+    }
+  }
+
+  handleInput(textElement.value, langElement.value, requestedAttributes, tokenizer);
+}
+
+/** Submits the input to Perspective and loads the appropriate output */
+async function handleInput(text, lang, requestedAttributes, tokenizer) {
+  // Draw the separating line for the output
+  const separator = document.getElementById('separator-container');
+  separator.innerHTML = '';
+  separator.appendChild(document.createElement('hr')); 
+	
+  // Make Perspective call for the entire submission and load graph data
+  const toxicityData = await callPerspective(text, lang, requestedAttributes);
+  loadChartsApi(toxicityData);
+
+  // Get detailed analysis if requested
+  if (tokenizer != "") {
+    getAnalysis(text, lang, requestedAttributes, tokenizer); 
+  }
+}
+
+/** Prints detailed analysis of text by word or sentence */
+async function getAnalysis(text, lang, requestedAttributes, tokenizer) {
+  // Set up the detailed analysis section
+  const analysisContainer = document.getElementById('analysis-container');
+  analysisContainer.appendChild(createAnyElement('b', 'Detailed Anaylsis'));
+  const loadingEl = document.createElement('p');
+  loadingEl.className = 'spinner-border';
+  analysisContainer.appendChild(loadingEl);
+  const result = document.createElement('p');
+  result.className = 'detailed-analysis';
+  
+  // Generate the results for every substring of the input text
+  const substrings = getSubstrings(text, tokenizer);
+  for (i = 0; i < substrings.length; i++) {
+    if (substrings[i] != '') {
+      // Get the Perspective scores for the substring & sort them descending
```

Comment:
It might be good to move (refactor) some of the code inside this `if` statement into it's own function, just to keep the `getAnalysis` function a bit more readable.

One option might be that you could still do `callPerspective` in the `getAnalysis` function, but then create a new function that updates the HTML elements for the response.

Also this isn't a big deal for now, but I believe the way `await` works in this case is that each `callPerspective` needs to be completed before the next one runs, which might slow down the UX a bit.  There are other ways you can write this, including using `Promise` or `Promise.all()`, where you can make all perspective requests in parallel, which may speed things up (you will have to be careful about variables used in the loop like `i` in that case).  This page has a little more info: https://stackoverflow.com/questions/11488014/asynchronous-process-inside-a-javascript-for-loop

Reply:
I followed your suggestion for refactoring the code. 

For the parallel calls - thank you! This works significantly faster.